### PR TITLE
AdminLTE 3 implementation

### LINKED ISSUE #5597: AdminLTE 3 implementation
## Feature Request

Hi, are you planned to implement AdminLTE 3? it looks better than 2 and seems modern

https://adminlte.io/themes/dev/AdminLTE/index.html

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

You can check result here:
[http://demo.wbloszyk.pl](http://demo.wbloszyk.pl)

I am targeting this branch, because it is BC-break.

Closes #5597

## Changelog

```markdown
### Added
- Added some `Class::newMethod` to do great stuff

### Changed

### Deprecated

### Removed

### Fixed

### Security
```

    
 ## To do    
- [x] Migrating to bootstrap v4 
- [x] Migrating to adminlte v3
- [x] Fix tests
- [ ] Fix Admin.js
    - [x] tab_errors
    - [ ] x-editable
- [ ] Check forms
    
## Update - 07.07.2020

Current assets system have many dependencies to other Sonata bundles (like `sonatacore` files). It mean it is not simple to upgrade vendors and templates. It is possible to do it at once at major update but require upgrade all sonata bundles at once. To avoid it and allow upgrade `sonata 3` to `sonata 4` one by one bundle this thigs will be do:
- new branch in sandbox where:
    - add encore webpack 
    - add bootstrap4 and AdminLTE3
    - prepare new templates 
- after finish this branch all this change will be split to bundles

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -25,9 +25,9 @@
         "php": "^7.3",
         "doctrine/common": "^2.7",
         "doctrine/inflector": "^1.1",
-        "knplabs/knp-menu": "^3.1",
-        "knplabs/knp-menu-bundle": "^3.0",
-        "sonata-project/block-bundle": "^4.0",
+        "knplabs/knp-menu": "^2.0 || ^3.1",
+        "knplabs/knp-menu-bundle": "^2.3 || ^3.0",
+        "sonata-project/block-bundle": "^3.18.3 || ^4.0",
```
Comment (VincentLanglet) [inline composer.json]: Why these changes ?

**Comment by VincentLanglet** [comment]
I think this will need an upgrade note for the user.

If he made some custom page, override, etc. He'll be happy to know which class to use instead of the previous one.

Something like,
Before / After
box -> card
box-primary -> card-primary
...

File: composer.json (hunk 1 above)
Comment (wbloszyk) [reply composer.json]: It is draft. I need it for working demo for now.

Comment (wbloszyk) [reply composer.json]: It is draft. I need it for working demo for now.

**Comment by wbloszyk** [comment]
I think refer for adminlte and bootstrap upgrade will be enought.

File: src/Resources/public/vendor/admin-lte/.babelrc [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (VincentLanglet) [inline src/Resources/public/vendor/admin-lte/.babelrc]: Do we have to commit all the admin-lte files ?
Can't we declare the dependency in the composer.json or something ?
Could we use https://packagist.org/packages/almasaeed2010/adminlte ?

File: composer.json (hunk 1 above)
Comment (VincentLanglet) [reply composer.json]: Ok, sorry for the too early review. 😅

**Comment by VincentLanglet** [comment]
> I think refer for adminlte and bootstrap upgrade will be enought.

Unfortunaly the AdminLTE release notes doesnt help a lot https://github.com/ColorlibHQ/AdminLTE/releases

And I didn't found the reason for your changes in https://getbootstrap.com/docs/4.0/migration/#stable-changes

Something more precise than `Update AdminLTE version from 2 to 3` could help the user.
When I took another look at your changes, and the only big change you made is actually
```
box box-primary
```
Into
```
card card-primary card-outline
```

What about adding a line in the upgrade note like 
- We're now using `card card-primary card-outline` instead of `box box-primary` ?

And if we makes some others class changes, adding a note for it.

**Comment by wbloszyk** [comment]
You can find adminlte upgrade guide here: [https://adminlte.io/docs/3.0/upgrade-guide.html](https://adminlte.io/docs/3.0/upgrade-guide.html)

I waiting for webpack. At first i will end "Drop SonataCoreBundle" PRs. At second "add support for BlockBundle 4.0. Then i will end this PR. 

BTW If adminlte and bootstrap upgrade guide is not clear enought i can add own upgrade notes.

**Comment by wbloszyk** [comment]
What with https://github.com/sonata-project/SonataAdminBundle/issues/6147 ??

If x-editable will be drop then I can finish this PR.

**Comment by jordisala1991** [comment]
IMHO https://github.com/sonata-project/SonataAdminBundle/pull/5461 this should be finish first, having to review 200k lines PR because we don't have assets correctly installed with npm is super hard and keep adding them makes that other PR harder and harder

**Comment by wbloszyk** [comment]
I dont want psuh assets but is require for tests.

**Comment by wbloszyk** [comment]
https://github.com/sonata-project/SonataAdminBundle/pull/6145#issuecomment-646545695

@greg0ire Can you merge 3.x branch to master in AdminBundle?

**Comment by martinbroos** [comment]
> You can find adminlte upgrade guide here: https://adminlte.io/docs/3.0/upgrade-guide.html
> 
> I waiting for webpack. At first i will end "Drop SonataCoreBundle" PRs. At second "add support for BlockBundle 4.0. Then i will end this PR.
> 
> BTW If adminlte and bootstrap upgrade guide is not clear enought i can add own upgrade notes.

I'm happy to help with implementing these upgrades defined in the upgrade guide of Admin LTE3. As I can see the implementation is still incomplete. For instance admin LTE3 requires font-awesome5 which uses 'fas' instead of 'fa' when using icons. 

Is it possible to ignore the webpack-encore update for now so we can get the theme implemented correctly. Others can help with this if we can get some of your code merged. If you want you can focus on 'Drop SonataCoreBundle' but my opinion is that we should not create a lot of dependencies between pull requests.

Could you fix the conflicts? Or is something else blocking this from merging ?

**Comment by wbloszyk** [comment]
The problem for now is x-editable not support bootstrap 4.

**Comment by martinbroos** [comment]
Ok thanks for your response. At the moment i'm not using x-editable and are using a custom theme. So hard to say if we should hold the release for this feature. We might be able to build something like this in combination with Popper.js which is part of admin lte 3 but it's hard to get it to work in way x-editable did with all those field types.
I'm not seeing any alternative library that provides this kind of functionality. 

But if the goal is using bootstrap 4 and admin lte3 my advice is to disable x-editable for now so other can help with finding a replacement or further implement this theme.

**Comment by VincentLanglet** [comment]
There is an issue for this https://github.com/sonata-project/SonataAdminBundle/issues/6147
x-editable is not maintained anymore...
